File: crates/strata-core/src/expr.rs
//! Safe arithmetic expression evaluator (Pratt parser).
//!
//! Mirrors `@strata/scene/expr.ts` — same grammar, same tests, same semantics.
//!
//! Grammar:
//!   expr     -> term (('+' | '-') term)*
//!   term     -> factor (('*' | '/') factor)*
//!   factor   -> NUMBER | ALIAS | '(' expr ')'
//!
//! Aliases: `{name}` resolved via a lookup map.
//! No unsafe code, no dynamic dispatch beyond the op table.
//!
//! Research basis: Pratt parsing (Vaughan Pratt 1973), same TS→Rust mirror
//! strategy as `strata-core::component`.

use std::collections::HashMap;

#[derive(Debug, Clone, PartialEq)]
enum Token {
    Number(f64),
    Alias(String),
    Op(char),
    Paren(char),
}

fn tokenize(input: &str) -> Result<Vec<Token>, String> {
    let mut tokens = Vec::new();
    let mut chars = input.chars().peekable();

    while let Some(&ch) = chars.peek() {
        if ch.is_whitespace() {
            chars.next();
            continue;
        }
        if ch == '{' {
            chars.next();
            let mut name = String::new();
            loop {
                match chars.next() {
                    None => return Err("Unclosed alias".into()),
                    Some('}') => break,
                    Some(c) => name.push(c),
                }
            }
            tokens.push(Token::Alias(name));
            continue;
        }
        if ch.is_ascii_digit() {
            let mut num = String::new();
            while let Some(&c) = chars.peek() {
                if c.is_ascii_digit() || c == '.' {
                    num.push(c);
                    chars.next();
                } else {
                    break;
                }
            }
            let val: f64 = num.parse().map_err(|_| format!("Invalid number: {num}"))?;
            tokens.push(Token::Number(val));
            continue;
        }
        if "+-*/".contains(ch) {
            chars.next();
            tokens.push(Token::Op(ch));
            continue;
        }
        if ch == '(' || ch == ')' {
            chars.next();
            tokens.push(Token::Paren(ch));
            continue;
        }
        return Err(format!("Unexpected character: '{ch}'"));
    }

    Ok(tokens)
}

fn precedence(op: char) -> u8 {
    match op {
        '+' | '-' => 1,
        '*' | '/' => 2,
        _ => 0,
    }
}

fn apply_op(op: char, a: f64, b: f64) -> Result<f64, String> {
    match op {
        '+' => Ok(a + b),
        '-' => Ok(a - b),
        '*' => Ok(a * b),
        '/' => {
            if b == 0.0 {
                Err("Division by zero".into())
            } else {
                Ok(a / b)
            }
        }
        _ => Err(format!("Unknown operator: '{op}'")),
    }
}

/// Evaluate an arithmetic expression string with alias resolution.
///
/// `aliases` provides resolved numeric values for `{name}` references.
/// Returns `Err` on syntax errors, unknown aliases, or division by zero.
pub fn evaluate(input: &str, aliases: &HashMap<String, f64>) -> Result<f64, String> {
    let tokens = tokenize(input)?;
    let mut pos = 0;
    let result = parse_expr(&tokens, &mut pos, 0, aliases)?;
    if pos < tokens.len() {
        return Err("Unexpected trailing tokens".to_string());
    }
    Ok(result)
}

fn parse_expr(
    tokens: &[Token],
    pos: &mut usize,
    min_prec: u8,
    aliases: &HashMap<String, f64>,
) -> Result<f64, String> {
    let mut left = parse_factor(tokens, pos, aliases)?;

    while let Some(Token::Op(c)) = tokens.get(*pos) {
        let op = *c;
        let prec = precedence(op);
        if prec < min_prec {
            break;
        }
        *pos += 1;
        let right = parse_expr(tokens, pos, prec + 1, aliases)?;
        left = apply_op(op, left, right)?;
    }

    Ok(left)
}

fn parse_factor(
    tokens: &[Token],
    pos: &mut usize,
    aliases: &HashMap<String, f64>,
) -> Result<f64, String> {
    let t = tokens
        .get(*pos)
        .ok_or_else(|| "Unexpected end of expression".to_string())?;
    *pos += 1;

    match t {
        Token::Number(n) => Ok(*n),
        Token::Alias(name) => aliases
            .get(name)
            .copied()
            .ok_or_else(|| format!("Unknown alias: {name}")),
        Token::Paren('(') => {
            let val = parse_expr(tokens, pos, 0, aliases)?;
            match tokens.get(*pos) {
                Some(Token::Paren(')')) => {
                    *pos += 1;
                    Ok(val)
                }
                _ => Err("Mismatched parentheses".into()),
            }
        }
        _ => Err(format!("Unexpected token: {t:?}")),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn aliases(pairs: &[(&str, f64)]) -> HashMap<String, f64> {
        pairs.iter().map(|(k, v)| (k.to_string(), *v)).collect()
    }

    #[test]
    fn evaluates_simple_number() {
        assert_eq!(evaluate("3", &HashMap::new()), Ok(3.0));
    }

    #[test]
    fn evaluates_addition() {
        assert_eq!(evaluate("1 + 2", &HashMap::new()), Ok(3.0));
    }

    #[test]
    fn evaluates_subtraction() {
        assert_eq!(evaluate("5 - 3", &HashMap::new()), Ok(2.0));
    }

    #[test]
    fn evaluates_multiplication() {
        assert_eq!(evaluate("3 * 4", &HashMap::new()), Ok(12.0));
    }

    #[test]
    fn evaluates_division() {
        assert_eq!(evaluate("10 / 2", &HashMap::new()), Ok(5.0));
    }

    #[test]
    fn respects_precedence() {
        assert_eq!(evaluate("2 + 3 * 4", &HashMap::new()), Ok(14.0));
    }

    #[test]
    fn respects_parentheses() {
        assert_eq!(evaluate("(2 + 3) * 4", &HashMap::new()), Ok(20.0));
    }

    #[test]
    fn chained_division_left_to_right() {
        assert_eq!(evaluate("8 / 4 / 2", &HashMap::new()), Ok(1.0));
    }

    #[test]
    fn chained_subtraction_left_to_right() {
        assert_eq!(evaluate("10 - 3 - 2", &HashMap::new()), Ok(5.0));
    }

    #[test]
    fn evaluates_alias_lookup() {
        let a = aliases(&[("base", 10.0)]);
        assert_eq!(evaluate("{base}", &a), Ok(10.0));
    }

    #[test]
    fn evaluates_alias_in_expression() {
        let a = aliases(&[("base", 10.0)]);
        assert_eq!(evaluate("{base} * 1.5", &a), Ok(15.0));
    }

    #[test]
    fn evaluates_alias_with_hyphen() {
        let a = aliases(&[("space-2", 8.0)]);
        assert_eq!(evaluate("{space-2} + 4", &a), Ok(12.0));
    }

    #[test]
    fn unknown_alias_returns_error() {
        let result = evaluate("{unknown}", &HashMap::new());
        assert!(result.is_err());
        assert!(result.unwrap_err().contains("Unknown alias"));
    }

    #[test]
    fn division_by_zero_returns_error() {
        let result = evaluate("1 / 0", &HashMap::new());
        assert!(result.is_err());
        assert!(result.unwrap_err().contains("Division by zero"));
    }

    #[test]
    fn malformed_expression_returns_error() {
        let result = evaluate("1 +", &HashMap::new());
        assert!(result.is_err());
    }

    #[test]
    fn mismatched_parentheses_returns_error() {
        let result = evaluate("(1 + 2", &HashMap::new());
        assert!(result.is_err());
    }

    #[test]
    fn invalid_character_returns_error() {
        let result = evaluate("2 @ 3", &HashMap::new());
        assert!(result.is_err());
    }
}

File: packages/scene/src/expr.ts
/**
 * Safe arithmetic expression evaluator (Pratt parser).
 *
 * Grammar:
 *   expr     -> term (('+' | '-') term)*
 *   term     -> factor (('*' | '/') factor)*
 *   factor   -> NUMBER | ALIAS | FUNC '(' args ')' | '(' expr ')'
 *   args     -> expr (',' expr)*
 *
 * Functions: min(…), max(…), round(…), ceil(…), floor(…)
 * Aliases: `{name}` — resolved via a lookup map.
 * No `eval()`, no `Function()`, no assignment.
 *
 * Research basis: Pratt parsing (Vaughan Pratt, 1973) — the standard approach
 * for safe expression evaluation without code generation.
 */

export type Token =
  | { kind: 'number'; value: number }
  | { kind: 'alias'; name: string }
  | { kind: 'op'; op: string }
  | { kind: 'paren'; value: string }
  | { kind: 'func'; name: string }
  | { kind: 'comma' };

type BinOp = (a: number, b: number) => number;

const PRECEDENCE: Record<string, number> = {
  '+': 1,
  '-': 1,
  '*': 2,
  '/': 2,
};

const BINARY_OPS: Record<string, BinOp> = {
  '+': (a, b) => a + b,
  '-': (a, b) => a - b,
  '*': (a, b) => a * b,
  '/': (a, b) => {
    if (b === 0) throw new Error('Division by zero');
    return a / b;
  },
};

const FUNCTIONS: Record<string, (...args: number[]) => number> = {
  min: (...args) => Math.min(...args),
  max: (...args) => Math.max(...args),
  round: (x) => Math.round(x),
  ceil: (x) => Math.ceil(x),
  floor: (x) => Math.floor(x),
};

export function tokenize(input: string): Token[] {
  const tokens: Token[] = [];
  let i = 0;
  while (i < input.length) {
    const ch = input[i];
    if (!ch) break;
    if (ch === ' ' || ch === '\t' || ch === '\n') {
      i++;
      continue;
    }
    if (ch === '{') {
      const end = input.indexOf('}', i);
      if (end < 0) throw new Error('Unclosed alias');
      tokens.push({ kind: 'alias', name: input.slice(i + 1, end) });
      i = end + 1;
      continue;
    }
    if (ch >= '0' && ch <= '9') {
      let j = i;
      while (j < input.length) {
        const cj = input[j];
        if (cj === undefined) break;
        if (!((cj >= '0' && cj <= '9') || cj === '.')) break;
        j++;
      }
      tokens.push({ kind: 'number', value: Number.parseFloat(input.slice(i, j)) });
      i = j;
      continue;
    }
    if (ch === ',') {
      tokens.push({ kind: 'comma' });
      i++;
      continue;
    }
    if ('+-*/'.includes(ch)) {
      tokens.push({ kind: 'op', op: ch });
      i++;
      continue;
    }
    if (ch === '(' || ch === ')') {
      tokens.push({ kind: 'paren', value: ch });
      i++;
      continue;
    }
    // Function names are alpha sequences followed by '('
    if ((ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z') || ch === '_') {
      let j = i;
      while (j < input.length) {
        const cj = input[j];
        if (cj === undefined) break;
        if (
          !(
            (cj >= 'a' && cj <= 'z') ||
            (cj >= 'A' && cj <= 'Z') ||
            (cj >= '0' && cj <= '9') ||
            cj === '_'
          )
        )
          break;
        j++;
      }
      const name = input.slice(i, j);
      // Check if followed by '(' - it's a function call
      // Skip whitespace to check
      let k = j;
      while (k < input.length && (input[k] === ' ' || input[k] === '\t' || input[k] === '\n')) k++;
      if (k < input.length && input[k] === '(') {
        tokens.push({ kind: 'func', name });
        i = j;
        continue;
      }
      throw new Error(`Unexpected identifier: '${name}'`);
    }
    throw new Error(`Unexpected character: '${ch}'`);
  }
  return tokens;
}

/**
 * Evaluate an arithmetic expression string with alias resolution.
 * `aliases` provides resolved numeric values for `{name}` references.
 * Throws on syntax errors, unknown aliases, or division by zero.
 */
export function evaluate(input: string, aliases: Record<string, number>): number {
  const tokens = tokenize(input);
  let pos = 0;

  function peek(): Token | null {
    return tokens[pos] ?? null;
  }

  function consume(): Token {
    const t = peek();
    if (!t) throw new Error('Unexpected end of expression');
    pos++;
    return t;
  }

  function parseExpr(minPrec: number): number {
    let left = parseFactor();

    while (true) {
      const t = peek();
      if (t?.kind !== 'op') break;
      const prec = PRECEDENCE[t.op] ?? 0;
      if (prec < minPrec) break;
      consume();
      const right = parseExpr(prec + 1);
      const op = BINARY_OPS[t.op];
      if (!op) throw new Error(`Unknown operator: ${t.op}`);
      left = op(left, right);
    }

    return left;
  }

  function parseArgs(): number[] {
    const args: number[] = [];
    // Check if no args (closing paren immediately)
    if (peek()?.kind === 'paren' && peek()?.value === ')') {
      return args;
    }
    args.push(parseExpr(0));
    while (peek()?.kind === 'comma') {
      consume(); // eat comma
      args.push(parseExpr(0));
    }
    return args;
  }

  function parseFactor(): number {
    // Unary minus
    if (peek()?.kind === 'op' && peek()?.op === '-') {
      consume(); // eat '-'
      const val = parseFactor();
      return -val;
    }
    const t = consume();
    if (t.kind === 'number') return t.value;
    if (t.kind === 'alias') {
      const val = aliases[t.name];
      if (val === undefined) throw new Error(`Unknown alias: ${t.name}`);
      return val;
    }
    if (t.kind === 'func') {
      const open = peek();
      if (open?.kind !== 'paren' || open.value !== '(') {
        throw new Error(`Expected '(' after function '${t.name}'`);
      }
      consume();
      const args = parseArgs();
      const close = peek();
      if (close?.kind !== 'paren' || close.value !== ')') {
        throw new Error('Mismatched parentheses in function call');
      }
      consume();
      const fn = FUNCTIONS[t.name];
      if (!fn) throw new Error(`Unknown function: ${t.name}`);
      return fn(...args);
    }
    if (t.kind === 'paren' && t.value === '(') {
      const val = parseExpr(0);
      const close = peek();
      if (close?.kind !== 'paren' || close.value !== ')') {
        throw new Error('Mismatched parentheses');
      }
      consume();
      return val;
    }
    throw new Error(`Unexpected token: ${JSON.stringify(t)}`);
  }

  const result = parseExpr(0);

  // Ensure no trailing tokens
  if (peek() !== null) {
    throw new Error('Unexpected trailing tokens');
  }

  return result;
}
